[Feature] Integrate `react-query` with PWA-Kit SDK

bendvc opened the pull request:
# Description

The objective of this PR is to integrate [React Query](https://tanstack.com/query/v4) with PWA-Kit, with the primary intention to allow use to build a hooks library that is compatible with our server-side rendering. Another motivating factor is to provide an alternative, not a replacement, to the aging `getProps` interface. 

This PR takes advantage of a library called `react-ssr-prepass` to record all usages of react-query in the "to be rendered" application so we can resolve those query requests, serialize the returned data, and continue on with the rending flow. 

The solution in this PR diverges slightly from the solution provided by @olibrook, in that we take special care to not serialize the awaiting of `getProps` and and `react-query` queries. In order to accomplish this I had to reorganize some of the code so that the _prepass_ stage happens before the "appState" is initialized. 

During the implementation I ran into a couple things I'd like some input on:

1. Because the `prepass` stage have been moved outside of the renderApp method, if it was to throw, we would have an uncaught error. Although I did some tests to get the prepass stage to fail, it seems like because it only traverses the JSX elements it doesn't throw errors itself if a component was to throw in its render. I'm guessing it's probably safer to ensure we catch errors no matter what and pass them along to the renderApp method. 
2. I've tried various methods, whether its setting defaultOptions on the `<Hydrate >` component so that or options on the queryClient, to stop hydrated queries from being called again. But I haven't found what would be an acceptable solution, aside from setting the `staleTime` to a value that would cause the query to _not_ be stale at the type of hydration. But that seems a little flaky. It would be nice to find a better solution for that.


_NOTE 1: The server side resolution of `react-query` queries is an opt-in feature, as the prepass state for rendering take proportional time to how long it takes to render. _

_NOTE 2: Support for dependent queries as described [here](https://tanstack.com/query/v4/docs/guides/dependent-queries) is currently NOT supported for server side rendering. Queries that are not enabled will not be run on the server._

## TODO's

- [ ] In order to provide a on-par solution to `getProps` we need to have access to the req/res objects. I'm thinking about creating a provider that will have the express request context. Maybe the name would be something like `useExpressCallbackArgs`. I'm open to name suggestions.
- [ ] Write tests.
- [ ] Handle `prepass` errors?
- [ ] Come to conclusion on handling how we stop queries from being triggered during hydration.


# Types of Changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] **Bug fix** (non-breaking change that fixes an issue)
- [ ] **New feature** (non-breaking change that adds functionality)
- [ ] **Documentation update**
- [ ] **Breaking change** (could cause existing functionality to not work as expected)
- [ ] **Other changes** (non-breaking changes that does not fit any of the above)

# Changes

- Added `ssrPrepassEnabled` feature flag. The reason for calling it `ssrPrepassEnabled` as opposed to something with react query or use query in it, is because you could still use useQuery client side, there would be no way to stop partners from doing that, so `enabledUseQuery` would be a misnomer. 

# How to Test-Drive This PR

-  Use the typescript minimal app. Its been set up to request data from a rando api endpoint that Oli set up. You can press the "Click Me" button to fetch a new Chuck Norris quote.

# Checklists

<!--- Enter an `x` in all the boxes that apply. -->
<!--- If you’re unsure about any of these, don’t hesitate to ask. We’re here to help! -->

## General

- [ ] Changes are covered by test cases
- [ ] CHANGELOG.md updated with a short description of changes (_not_ required for documentation updates)

Commit:
Demo react-query fetching w/o hard-coded keys on the server-side.

Commit:
Actually render fetched data

bendvc committed:
Server dehydration & client hydration

bendvc committed:
Integrate `react-query` with PWA-Kit SDK

bendvc committed:
Merge branch 'develop' into react-query-ssr-investigation

bendvc committed:
Rough-in the `useExpress` hook

bendvc committed:
Fix failing tests

kevinxh commented on packages/pwa-kit-react-sdk/package.json:
```diff
@@ -47,6 +47,7 @@
     "@loadable/babel-plugin": "^5.13.2",
     "@loadable/server": "^5.15.0",
     "@loadable/webpack-plugin": "^5.15.0",
+    "@tanstack/react-query": "^4.0.10",
```
Is it our intention to have `react-query` as a dependency, not peerDependency?

bendvc replied:
I think that we are in agreement that we don't want to make a breaking change right now. So we have to add this library as a dependency. I don't think it would hurt to add it as an optional peer dependency as well, that would position us for future iterations. What are your thoughts on this?

kevinxh commented on packages/template-typescript-minimal/app/pages/home.tsx:
```diff
@@ -83,12 +82,21 @@ h1 {
-        }, 1000)
-        return () => clearInterval(interval)
-    }, [counter, setCounter])
+    const {res: appRes} = useExpress()
+    const query = useQuery(['my-query', counter], async () => {
+        const res = await fetch(`https://api.chucknorris.io/jokes/random?couter=${counter}`)
+        if (appRes) {
+            appRes.status(404)
+        }
+        return await res.json()
+    })
```
If I understand this correctly, `useExpress` is a new SDK API that provide access to req and res on the server side.

On the client side, res and req will be undefined, right?

nit - Reading this code block, for a second there I thought we serialize and hydrate `req` and `res` on the client side. What if we name this thing `useServerContext` or something like that (include `server` in the name). So we make this very clear this is only applicable on the server side.

bendvc replied:
Yes that is what is does. I'm open to a name change. Let me run with that and see how it feels

kevinxh commented on packages/pwa-kit-react-sdk/src/ssr/server/react-rendering.js:
```diff
@@ -84,23 +87,27 @@ const initAppState = async ({App, component, match, route, req, res, location})
-              })
-            : Promise.resolve({})
-    )
+    const queries = queryClient.queryCache.queries.map((q) => q.fetch())
```
what happens when the query throw error? Is react-query able to hydrate the error?

bendvc replied:
There are some [limitations](https://tanstack.com/query/v4/docs/reference/hydration#limitations) with serialization of errors. I can play around with it to see the actual behaviour, because it's not completely clear what happens to errors if you tell dehydrate to serialize them. Since it says you can't do that.

kevinxh replied:
I see. From the documentation, it looks like react-query does not provide a serializer for errors.

bendvc replied:
It would be likely that the error would be handled one of a handful of ways that don't result in the need for serializing the error. 

1. The error is thrown and our server pipeline will do the appropriate things with it, e.g. set the status code, render the error page.
2. The error is not thrown and handled on the page in which the page would not render an appropriate UI. 

I still want to see what would happen is a query throws an error internally in its query function, then on the client we wouldn't re-run that query. I'm curious as to what data and or error are set to.

bendvc committed:
Catch prepass errors

vmarta commented on packages/pwa-kit-react-sdk/src/ssr/server/react-rendering.js:
```diff
@@ -208,29 +232,80 @@ export const render = async (req, res, next) => {
+
+    return (
+        <ExpressContext.Provider value={{req, res}}>
+            <QueryClientProvider client={queryClient}>
```
```suggestion
            <QueryClientProvider client={queryClient} contextSharing={true}>
```

In response to @kevinxh 's question on Slack recently: I think both `pwa-kit-react-sdk` and `commerce-sdk-react` packages can mount their own QueryClientProvider, that is, if `commerce-sdk-react` is meant to be a standalone library. By setting `contextSharing` to true, it looks like it's ok to have multiple QueryClientProvider in the component tree, and only one instance will be used. 

What `contextSharing` option does: https://tanstack.com/query/v4/docs/reference/QueryClientProvider

bendvc replied:
`commerce-sdk-react` isn't going to be mounting it's own provider. That being said, it doesn't mean that someone wouldn't add their own for some weird reason. The question is, do we want to make that decision for the user.. also does context only get shared amongst providers that have that value set to true? hmm

bendvc committed:
Remove unused query client config

bendvc committed:
Maybe a fix for the issue not finding react query

bendvc committed:
Clean up typescript min project.. fix webpack config.

bendvc committed:
Fix unit tests and coverage

bendvc committed:
Merge branch 'develop' into react-query-ssr-investigation

bendvc committed:
Undo changes to webpack

bendvc committed:
Merge branch 'react-query-ssr-investigation' of https://github.com/SalesforceCommerceCloud/pwa-kit into react-query-ssr-investigation

bendvc committed:
Add original alias back into webpack config

bendvc committed:
Ensure we don't serialize disabled queries...

this would waste space in the html.

alexvuong commented on packages/pwa-kit-react-sdk/src/ssr/server/react-rendering.js:
```diff
@@ -208,29 +235,79 @@ export const render = async (req, res, next) => {
     }
 }
```
@bendvc, I looked in the react-query guideline, and they have a section about clearing cache on the server, do we need to do that in our case? https://tanstack.com/query/v4/docs/guides/ssr#high-memory-consumption-on-server

bendvc replied:
Thanks for pointing that out. I think if you read the comment carefully you'll see that you actually don't need to call any clear cache methods, as long as you haven't changed the `cacheTime` value. I attached the excerpt below. 

`
On the server, cacheTime defaults to Infinity which disables manual garbage collection and will automatically clear memory once a request has finished. If you are explicitly setting a non-Infinity cacheTime then you will be responsible for clearing the cache early.
`

alexvuong commented on packages/pwa-kit-react-sdk/src/ssr/server/react-rendering.js:
```diff
@@ -83,24 +86,31 @@ const initAppState = async ({App, component, match, route, req, res, location})
 
+    const queryCache = queryClient.getQueryCache()
+    const queries = queryCache.getAll()
+    const queryPromises = queries.map(({fetch, enabled}) => enabled && fetch())
```
Thanks for getting this fixed quickly.
To recap a conversation that @bendvc and I have earlier, I raised a concern how the previous version `const queries = queryClient.query cache.queries.map((q) => q.fetch())` worked when the users do a dependent query calls. (e.g [here](https://tanstack.com/query/v4/docs/guides/dependent-queries)).  
In other words, it is possible that we can tell useQuery to not fetch until a  piece of information is available, and that is not possible previously. Hence it is necessary that we take the `enabled` into consideration here.

bendvc replied:
Thanks Alex. I have gon ahead and updated the code to 1. not fetch queries that aren't enabled in order to not bloat the html response and 2. use the react query api methods to get those queries and not accessing the possibly private members like **queryCache** and **queries**.

bendvc committed:
Fix conditional fetching of queries.

- Make useQuery example in ts minimal server side only

bendvc committed:
Update CHANGELOG.md

bendvc committed:
Typo fix

bendvc committed:
Playing with adding react-query via hoc

bendvc committed:
Commenting and minor changes to function names

bendvc committed:
Fix hoc logic not working with retail template

bendvc committed:
Fix linting

olibrook commented:
@bendvc I took a look at your HOC work. This is definitely the direction I was imagining. I spent some time tinkering on Friday and didn't quite get anything presentable. Big picture, I'd suggest some acceptance criteria for you first:


1. Let's agree that we are fully replacing our _recommendation_ to use `getProps` with a recommendation to use `react-query` for all pwa-kit data-fetching, moving forward. Design for it as the new default.
2. Existing `getProps` users should not pay performance penalties for `react-query` if they're not using it (neither bundle size, nor SSR speed).
3. We don't want to invent new config for this that we think we'll come to regret (eg. `ssrOnly` or the proposed `ssrPrepass`).

If we can agree on that, we can talk APIs! Fundamentally, I think we need to move the data-fetching parts out of the react-rendering code and into something that the user can configure. So, [where we have this] in `render`:

```js
    // Step 3 - Init the app state
    const {appState, error: appStateError} = await initAppState({
        App: WrappedApp,
        component,
        match,
        route,
        req,
        res,
        location
    })
```

We should find a way to do this instead, handing responsibility to the `App` or a HOC to do the `initAppState` stuff:

```js
    const {appState, error: appStateError} = await WrappedApp.initAppState({
        App: WrappedApp,
        component,
        match,
        route,
        req,
        res,
        location
    })
```

You could then find a way of making the `WrappedApp` contents swappable, and using an HOC sounds like a great way of doing it. This is nearly what you have, except I'm suggesting offload `initAppState` to the HOC in its entirety, not just `getAllQueryPromises`.

Once we've done that, I would strongly recommend that we make the `react-query` behaviour the default with no additional HOC at all in user-space. I would set it up in such a way that `getProps` users need to *opt-out* of the new default, by adding an HOC to an App component. We want the burden of config for people that are using `react-query` to be zero, if it's our recommended default.

So, I'd expect us to make a major version bump with a single breaking change: Existing `getProps` users that don't want to upgrade have to opt out of the new behaviour by doing something like this in user code:

```
export default useLegacyGetProps(App)
```

This _is_ a breaking change, but very easy to do for users who don't want to upgrade their code. Perfect example of when I think we should do it!


[where we have this]: https://github.com/SalesforceCommerceCloud/pwa-kit/blob/develop/packages/pwa-kit-react-sdk/src/ssr/server/react-rendering.js#L162

bendvc commented:
> 

I like the idea of offloading the initAppState function. That makes it very easy to define what the API is for any of those HOC's. I'll play around with it and let you know if I have any questions. 

UPDATE: As per our discussion on friday, we do want to ensure that react-query can in fact do everything getProps does if we are going to make it our recommended default in the major version bump. Unless you are recommending the API allow you to ultimately stack these data providing HOC's on top of each other, but I don't think you are as that would probably get quite complex pretty fast.

bendvc committed:
Split up API functionalities, add rough stacking of API functions

olibrook commented:
@bendvc @johnboxall People are rightly worried about breaking changes. I think that the kind we should care about are those that have no justification and no benefit to the end user. Is this one of those occasions?

My answer is no. You haven't spelled it out, but I think you are going to re-write the retail-react template in the near future so that it barely uses `getProps`, if at all. When you do that, I think you're going to find that when users ask _"how should I fetch data from [service XYZ] in my app?"_, you'll want to recommend `use-query` – a single answer, across the board.

As such, I think you should bias your design towards what you are going to be recommending users do by default soon.

If you do a major version bump then your upgrade guide is going to say something like this:

> We've added support for react-query in this release. If you are happy using `getProps` and you want to opt-out of 
> react-query support, wrap your `App` component with the `useLegacyGetProps` HOC – if you do this, you'll avoid doing an
> SSR prepass on the server-side, which will improve your performance.

This would be a very easy upgrade guide to follow and would be balanced against the – apparent! - benefit of switching to something hooks-based, which y'all clearly want to do...

bendvc committed:
Update react-rendering.js

bendvc committed:
Clean up.. Make getProps the default... Etc

bendvc committed:
Clean up function and file names

alexvuong commented on packages/pwa-kit-react-sdk/src/ssr/server/react-rendering.js:
```diff
@@ -162,30 +150,56 @@ export const render = async (req, res, next) => {
+            routes
+        })
+
+        ;({appState, error: appStateError} = await WrappedApp.fetchState({
```
Dumb question, why do we need to wrap this line around `;( ... )` ?

bendvc replied:
This line isn't in the code anymore, but for education purposes. I found out that this is how you accomplish using let to declare variables and have them be assigned to during destructuring. 

The reason I was having to do this is because `appState` and `appStateError` were declared outside of this conditional because they are being used elsewhere. Alternatively you could use an intermediate variable like this

```
const result = await WrappedApp.fetchState(...)
appState = result.appState
appStateError = result.appStateError
```

bendvc committed:
Change API function name. Change to return promises.

bendvc committed:
Refactor routeComponent into withLegaryGetProps, refactor withErrorHandling

bendvc committed:
move hoc's into component folder

bendvc committed:
Fixed tests for withLegacyGetProps

bendvc committed:
Small function name fixes

bendvc committed:
Clean up.. linting.. comments.. etc

bendvc committed:
Add tests for with react query hoc

bendvc committed:
Fix error handling distinction between render, state, etc

vmarta commented:
@bendvc Follow-up to our earlier conversation on Slack: yup, there is a bug that can be seen by loading the retail-react-app-template, specifically with how `extraGetPropsArgs` is being shared across components. 

In the retail-react-app:
- `api` is undefined in the `App.getProps`
- BUT it is defined in page components like `Home.getProps`

bendvc committed:
All getProps tests are passing now.

bendvc committed:
Re-add useQuery tests

vmarta commented on packages/pwa-kit-react-sdk/src/ssr/server/react-rendering.js:
```diff
@@ -162,30 +127,74 @@ export const render = async (req, res, next) => {
+                {}
+            )
+        } catch (e) {
+            appStateError = logAndFormatError(e || new Error())
```
Querying Commerce API for a non-existent category, for example, would throw an error, which would get caught in this catch statement. The end result is this internal server error:

![Google Chrome 2022-08-30 at 16 50 48](https://user-images.githubusercontent.com/847300/187563315-d5da5172-973b-44a5-8259-8c6b88101c9b.png)



But what if the developers would want to catch and handle it in their app? Currently they can't because it's caught in the react rendering.

vmarta replied:
One scenario I'm investigating is for the useExpress hook: how to detect a non-existent category and spit out a 404. I would want to be able to do the following in the user land:

```jsx
    const {categoryId} = useParams()
    const {data, isLoading, error} = useCategory({id: categoryId, levels: 0})
    const {req, res: pageRes} = useExpress()

    if (pageRes && error?.response?.status === 404) {
        pageRes.status(404).send(data?.detail)
    }
```

We were able to do a similar thing in `getProps`: https://github.com/SalesforceCommerceCloud/pwa-kit/blob/59a165fca4e175133b3f18ea6d1b256b272439e7/packages/template-retail-react-app/app/pages/product-list/index.jsx#L599-L603

bendvc committed:
Fix errors thrown by withReactRouter data fetching method

bendvc committed:
Fix retail template to use legacy getProps

bendvc committed:
Merge branch 'develop' into react-query-ssr-investigation

bendvc committed:
Re-build lock files

vmarta commented on packages/pwa-kit-react-sdk/src/ssr/universal/components/with-react-query/index.jsx:
```diff
@@ -0,0 +1,90 @@
+                const queries = queryCache.getAll()
+                const promises = queries
+                    .filter(({options}) => options.enabled !== false)
+                    .map((query) => query.fetch().catch((e) => ({error: e})))
```
In case users would want to use the other error-related properties that are returned by `useQuery`:

```suggestion
                    .map((query) => query.fetch().catch((e) => ({error: e, status: 'error', isError: true})))
```

bendvc committed:
Refactor

Split useLegarcygetProps and routeComponent again
Move logic for setting preloadedProps into the HOC
Fix issue with useQuery throwing in render

bendvc committed:
Testing new initStaticContext api

vmarta commented on packages/pwa-kit-react-sdk/src/ssr/universal/components/with-react-query/index.jsx:
```diff
@@ -0,0 +1,108 @@
+                    .map((query) => query.fetch().catch((error) => {
+                        // NOTE: Our best attempt to create a logical return object without
+                        // getting out of hand.
+                        return {
+                            error,
+                            errorUpdatedAt: Date.now(),
+                            errorUpdateCount: 1,
+                            isError: true,
+                            isLoadingError: true,
+                            isRefetchError: false,
+                            status: 'error'
+                        }
```
I found that this object is _not_ what I see on the user land. The object has more things in it.. more complete.

Is this because although we're returning all promises on line 82 below, what gets saved is actually the `queryClient` itself on line 84?

bendvc committed:
Merge branch 'develop' into react-query-ssr-investigation
